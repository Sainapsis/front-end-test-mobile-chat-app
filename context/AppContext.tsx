import React, { createContext, useContext, ReactNode } from 'react';
import { useUser } from '@/hooks/useUser';
import { useChats } from '@/hooks/useChats';
import { DatabaseProvider } from '../database/DatabaseProvider';
import { useDatabase } from '@/hooks/useDatabase';
import { User } from '@/types/User';
import { Chat } from '@/types/Chat';

/**
 * Interface representing the application context state
 */
type AppContextType = {
  users: User[];
  currentUser: User | null;
  isLoggedIn: boolean;
  login: (userId: string) => Promise<boolean>;
  logout: () => void;
  /** List of chats for the current user */
  chats: Chat[];
  createChat: (participantIds: string[]) => Promise<Chat | null>;
  sendMessage: (chatId: string, text: string, senderId: string, imageUri?: string) => Promise<boolean>;
  loading: boolean;
  dbInitialized: boolean;
  clearChats: (userId: string) => Promise<void>;
  deleteChat?: (chatId: string, userId: string) => Promise<void>;
  deleteMessage?: (messageId: string, chatId: string) => Promise<boolean>;
  addReaction?: (messageId: string, emoji: string) => Promise<boolean>;
  removeReaction?: (reactionId: string, messageId: string) => Promise<boolean>;
  editMessage?: (messageId: string, newText: string) => Promise<boolean>;
};

/**
 * Context for managing application-wide state
 */
const AppContext = createContext<AppContextType | undefined>(undefined);

/**
 * Internal component that wraps the application content with context
 * @param children - Child components to be wrapped by the context
 */
function AppContent({ children }: { children: ReactNode }) {
  const { isInitialized } = useDatabase();
  const userContext = useUser();
  const chatContext = useChats(userContext.currentUser?.id || null);
  
  const loading = !isInitialized || userContext.loading || chatContext.loading;

  const value: AppContextType = {
    ...userContext,
    ...chatContext,
    loading,
    dbInitialized: isInitialized,
    deleteChat: chatContext.deleteChat || undefined,
    deleteMessage: chatContext.deleteMessage || undefined,
    addReaction: chatContext.addReaction || undefined,
    removeReaction: chatContext.removeReaction || undefined,
    editMessage: chatContext.editMessage || undefined,
    sendMessage: async (chatId: string, text: string, senderId: string, imageUri?: string) => {
      if (!chatContext.sendMessage) return false;
      return chatContext.sendMessage(chatId, text, senderId, imageUri);
    },
  };

  if (!isInitialized) {
    return null;
  }

  return <AppContext.Provider value={value}>{children}</AppContext.Provider>;
}

/**
 * Provider component that wraps the application with necessary contexts
 * @param children - Child components to be wrapped by the provider
 */
export const AppProvider = ({ children }: { children: React.ReactNode }) => (
  <DatabaseProvider>
    <AppContent>{children}</AppContent>
  </DatabaseProvider>
);

/**
 * Hook for accessing application context
 * @returns Application context containing user, chat, and database states
 * @throws Error if used outside of AppProvider
 */
export function useAppContext() {
  const context = useContext(AppContext);
  if (!context) {
    throw new Error('useAppContext must be used within an AppProvider');
  }
  return context;
}
