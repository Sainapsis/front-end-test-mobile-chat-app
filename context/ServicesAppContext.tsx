import React, { createContext, useContext, ReactNode } from 'react';
import { DatabaseProvider, useDatabaseStatus } from './DataBaseContext';
import { useChatsDb,  useUserDb } from '../hooks/db';
import { Chat, Message, User } from '@/utils/types';

type AppContextType = {
  users: User[];
  currentUser: User | null;
  isLoggedIn: boolean;
  login: (userId: string) => Promise<boolean>;
  logout: () => void;
  chats: Chat[];
  createChat: (participantIds: string[]) => Promise<Chat | null>;
  sendMessage: (
    chatId: string, 
    text: string | undefined, 
    senderId: string,
    mediaUrl?: string,
    mediaType?: string,
    mediaThumbnail?: string,
    voiceUrl?: string,
    voiceDuration?: number,
    isVoiceMessage?: boolean,
    
  ) => Promise<Message | null>;
  addReaction: (messageId: string, userId: string, emoji: string) => Promise<boolean>;
  removeReaction: (reactionId: string) => Promise<boolean>;
  searchMessages: (query: string, chatId?: string) => Promise<void>;
  searchResults: Message[];
  deleteMessage: (messageId: string) => Promise<boolean>;
  editMessage: (messageId: string, newText: string) => Promise<boolean>;
  updateMessageStatus: (messageId: string, status: 'sent' | 'delivered' | 'read') => Promise<void>;
  markMessagesAsRead: (chatId: string) => Promise<void>;
  loading: boolean;
  dbInitialized: boolean;
};

export const AppContext = createContext<AppContextType | undefined>(undefined);

function AppContent({ children }: { children: ReactNode }) {
  const { isInitialized } = useDatabaseStatus();
  const userContext = useUserDb();
  const chatContext = useChatsDb(userContext.currentUser?.id || null);
  
  const loading = !isInitialized || userContext.loading || chatContext.loading;

  const value = {
    ...userContext,
    ...chatContext,
    loading,
    dbInitialized: isInitialized,
  };

  return <AppContext.Provider value={value}>{children}</AppContext.Provider>;
}

export function AppProvider({ children }: { children: ReactNode }) {
  return (
    <DatabaseProvider>
      <AppContent>{children}</AppContent>
    </DatabaseProvider>
  );
}

