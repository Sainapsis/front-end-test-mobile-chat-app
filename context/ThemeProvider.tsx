import React, { useState, useEffect } from 'react';
import AsyncStorage from '@react-native-async-storage/async-storage';
import ThemeContext from './ThemeContext';
import { Theme } from '@/types/tColores';

/**
 * Provider component that handles theme state and persistence
 * @param children - Child components to be wrapped by the provider
 */
export const ThemeProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  const [theme, setTheme] = useState<Theme>('light');

  useEffect(() => {
    loadTheme();
  }, []);

  const loadTheme = async () => {
    try {
      const savedTheme = await AsyncStorage.getItem('theme');
      if (savedTheme) {
        setTheme(savedTheme as Theme);
      }
    } catch (error) {
      console.error('Error loading theme:', error);
    }
  };

  const toggleTheme = async () => {
    const newTheme = theme === 'light' ? 'dark' : 'light';
    setTheme(newTheme);
    try {
      await AsyncStorage.setItem('theme', newTheme);
    } catch (error) {
      console.error('Error saving theme:', error);
    }
  };

  return (
    <ThemeContext.Provider value={{ theme, toggleTheme }}>
      {children}
    </ThemeContext.Provider>
  );
};