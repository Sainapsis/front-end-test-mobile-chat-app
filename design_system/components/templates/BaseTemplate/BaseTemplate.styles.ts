import { StyleSheet } from 'react-native';

export const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  header: {
    zIndex: 1,
  },
  content: {
    flex: 1,
  },
  footer: {
    borderTopWidth: 1,
  }
});