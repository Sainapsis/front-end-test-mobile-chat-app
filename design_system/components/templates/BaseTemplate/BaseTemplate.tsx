import React from 'react';
import { StatusBar, View, ViewStyle, StyleProp } from 'react-native';
import { styles } from './BaseTemplate.styles';
import { ThemedView } from '@/design_system/components/atoms';

interface BaseTemplateProps {
  /** Main content of the template */
  children: React.ReactNode;
  /** Optional header component */
  header?: React.ReactNode;
  /** Optional footer component */
  footer?: React.ReactNode;
  /** Custom styles to be applied to the template */
  style?: StyleProp<ViewStyle>;
  /** Test ID for testing purposes */
  testID?: string;
}

/**
 * BaseTemplate component provides a basic layout structure for screens.
 * It includes optional header and footer sections, and supports custom styling.
 */
export const BaseTemplate: React.FC<BaseTemplateProps> = ({
  children,
  header,
  footer,
  style,
}) => (
  <ThemedView testID="base-template" style={[styles.container, style]}>
    <StatusBar barStyle="dark-content" />
    {header && <View style={styles.header}>{header}</View>}
    <View style={styles.content}>{children}</View>
    {footer && <View style={styles.footer}>{footer}</View>}
  </ThemedView>
);
