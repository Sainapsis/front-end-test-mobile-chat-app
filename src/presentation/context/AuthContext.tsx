import { LoginParams } from "@/src/data/interfaces/auth.interface";
import { createContext, ReactNode, useContext } from "react";
import { useAuth } from "../hooks/useAuth";
import { User } from "@/src/domain/entities/user";

interface AuthContextType {
  currentUser: User | null;
  login: ({ userId }: LoginParams) => Promise<boolean>;
  logout: () => void;
}

const _AuthContext = createContext<AuthContextType | undefined>(undefined);

export function AuthContext({ children }: { children: ReactNode }) {
  const authContext = useAuth();

  const value = {
    ...authContext,
  };

  return (
    <_AuthContext.Provider value={value}>{children}</_AuthContext.Provider>
  );
}

export function useAuthContext() {
  const context = useContext(_AuthContext);

  if (context === undefined) {
    throw new Error("useAuthContext must be used within an AppProvider");
  }

  return context;
}
