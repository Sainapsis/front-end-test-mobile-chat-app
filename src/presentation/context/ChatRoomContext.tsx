import { DeleteMessageParams, EditMessageParams, UpdateStatusMessageParams } from '@/src/data/interfaces/chatRoom.interface';
import { Chat } from '@/src/domain/entities/chat';
import { Message, MessageStatus } from '@/src/domain/entities/message';
import { createContext, ReactNode, useContext } from "react";
import { useChatRoom } from '../hooks/useChatRoom';

interface ChatRoomContextType {
  updateStatus: (
    currentUserId: string,
    chat: Chat,
    statusMessage: MessageStatus
  ) => Promise<void>;
  sendMessage: (chatId: string, message: Message) => Promise<boolean>;
  editMessage: ({
    chatId,
    messageId,
    newText,
  }: EditMessageParams) => Promise<boolean>;
  deleteMessage: ({
    chatId,
    messageId,
  }: DeleteMessageParams) => Promise<boolean>;
  updateMessageStatus: (
    chatId: string,
    { messageId, status }: UpdateStatusMessageParams
  ) => Promise<void>;
  handleLoadMoreMessage: ({ chatId }: { chatId: string }) => Promise<void>;
}

const _ChatRoomContext = createContext<ChatRoomContextType | undefined>(undefined);

export function ChatRoomContext({ chatId, children }: { chatId: string; children: ReactNode }) {
  const chatRoomContext = useChatRoom({ chatId });

  const value = {
    ...chatRoomContext,
  };

  return (
    <_ChatRoomContext.Provider value={value}>{children}</_ChatRoomContext.Provider>
  );
}

export function useChatRoomContext() {
  const context = useContext(_ChatRoomContext);

  if (context === undefined) {
    throw new Error("useChatRoomContext must be used within an AppProvider");
  }

  return context;
}
