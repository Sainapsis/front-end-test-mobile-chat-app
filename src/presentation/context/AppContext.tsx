import React, { createContext, useContext, ReactNode } from "react";
import { useDatabase } from "../hooks/useDatabase";
import { useUser } from "@/src/presentation/hooks/useUser";
import { DatabaseContext } from "@/src/presentation/context/DatabaseProvider";
import { UserContext } from "./UserContext";
import { AuthContext } from "./AuthContext";

interface AppContextType {
  dbInitialized: boolean;
}

const AppContext = createContext<AppContextType | undefined>(undefined);

function AppContent({ children }: { children: ReactNode }) {
  const { isInitialized } = useDatabase();
  const userContext = useUser();
  
  const value = {
    ...userContext,
    dbInitialized: isInitialized,
  };

  return <AppContext.Provider value={value}>{children}</AppContext.Provider>;
}

export function AppProvider({ chatId, children }: { chatId: string; children: ReactNode }) {
  return (
    <DatabaseContext>
      <AuthContext>
        <UserContext>
          <AppContent>{children}</AppContent>
        </UserContext>
      </AuthContext>
    </DatabaseContext>
  );
}

export function useAppContext() {
  const context = useContext(AppContext);

  if (context === undefined) {
    throw new Error("useAppContext must be used within an AppProvider");
  }

  return context;
}
