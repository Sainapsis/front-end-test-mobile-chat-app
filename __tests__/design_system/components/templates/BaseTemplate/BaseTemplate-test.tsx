import * as React from 'react';
import { render } from '@testing-library/react-native';
import renderer from 'react-test-renderer';
import { ThemeProvider } from '@/context/ThemeContext';
import { BaseTemplate } from '@/design_system/components/templates/BaseTemplate/BaseTemplate';
import { Text, View } from 'react-native';

// Mock para ThemedView
jest.mock('@/design_system/components/atoms/ThemedView', () => {
    const React = require('react');
    const { View } = require('react-native');
    return {
      ThemedView: ({ children, testID, style }: { children: React.ReactNode, testID?: string, style?: object }) => (
        <View testID={testID} style={style}>
          {children}
        </View>
      ),
    };
  });
  
  
  

describe('BaseTemplate Component', () => {
  const renderWithTheme = (component: React.ReactNode) => {
    return render(
      <ThemeProvider>
        {component}
      </ThemeProvider>
    );
  };

  it('renders correctly with snapshot', () => {
    const tree = renderer.create(
      <ThemeProvider>
        <BaseTemplate>
          <Text>Content</Text>
        </BaseTemplate>
      </ThemeProvider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders children correctly', () => {
    const { getByText } = renderWithTheme(
      <BaseTemplate>
        <Text>Test Content</Text>
      </BaseTemplate>
    );
    
    expect(getByText('Test Content')).toBeTruthy();
  });

  it('renders header when provided', () => {
    const { getByText } = renderWithTheme(
      <BaseTemplate
        header={<Text>Header Content</Text>}
      >
        <Text>Main Content</Text>
      </BaseTemplate>
    );
    
    expect(getByText('Header Content')).toBeTruthy();
    expect(getByText('Main Content')).toBeTruthy();
  });

  it('renders footer when provided', () => {
    const { getByText } = renderWithTheme(
      <BaseTemplate
        footer={<Text>Footer Content</Text>}
      >
        <Text>Main Content</Text>
      </BaseTemplate>
    );
    
    expect(getByText('Footer Content')).toBeTruthy();
    expect(getByText('Main Content')).toBeTruthy();
  });

  it('renders both header and footer when provided', () => {
    const { getByText } = renderWithTheme(
      <BaseTemplate
        header={<Text>Header Content</Text>}
        footer={<Text>Footer Content</Text>}
      >
        <Text>Main Content</Text>
      </BaseTemplate>
    );
    
    expect(getByText('Header Content')).toBeTruthy();
    expect(getByText('Main Content')).toBeTruthy();
    expect(getByText('Footer Content')).toBeTruthy();
  });

  it('applies custom styles when provided', () => {
    const customStyle = { backgroundColor: 'red' };
    const { getByTestId } = renderWithTheme(
      <BaseTemplate style={customStyle}>
        <Text>Styled Content</Text>
      </BaseTemplate>
    );
  
    // Verificamos que el componente con testID="base-template" est√© presente
    const container = getByTestId('base-template');
    expect(container).toBeTruthy();
  });
  
  
  
  
});
