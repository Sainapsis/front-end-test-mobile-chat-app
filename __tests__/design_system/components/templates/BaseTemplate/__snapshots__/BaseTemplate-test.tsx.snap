// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseTemplate Component renders correctly with snapshot 1`] = `
<View
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
  testID="base-template"
>
  <View
    style={
      {
        "flex": 1,
      }
    }
  >
    <Text>
      Content
    </Text>
  </View>
</View>
`;
